### PR TITLE
Adding a wrapper for exec calls.

### PR DESCRIPTION
**Pull request by adarsh-zededa**
Issue:
When there is a memory shortage on the device, few agents like zedagentmetric and zedrouter were stuck on exec calls.
Meanwhile, the service's touch file was not updated while the services were waiting for the exec call to finish. This eventually led to Watchdog reboots.

Fix:
Added a wrapper on exec call which along with the command, also accepts the agentName and while waiting for the exec call to finish, the wrapper method will keep punching the agent's touch file. 

Rearranged few method so that caller and callee methods are in the same file. This was needed to help with figuring out zedagent's subagent name from the stacktrace.

### CODE REVIEW

**Comment by rvs** [comment]
Hey @adarsh-zededa -- I like the overall approach, but the fact that you need to pass agent name into just about everything now seems quite annoying. How about we actually use https://golang.org/pkg/runtime/debug/#Stack and see which agent's stack we're running on -- whatever is at the bottom of the stack -- we shall touch the file.

This will cut down on a huge number of changes in the PR you need to make and get it to a pretty small size. The reason I actually care about it being small is because I think we need to re-consider our general approach to these touch files to begin with. The right approach given our architecture is a liveness check anyway -- so I'd rather focus on moving us there (hence refactoring this PR relatively quickly).

Also, it seems all the periodic_sigusr1.sh belongs to a different PR -- can you please split it out?

**Comment by adarsh-zededa** [comment]
> Hey @adarsh-zededa -- I like the overall approach, but the fact that you need to pass agent name into just about everything now seems quite annoying. How about we actually use https://golang.org/pkg/runtime/debug/#Stack and see which agent's stack we're running on -- whatever is at the bottom of the stack -- we shall touch the file.

Done.

> Also, it seems all the periodic_sigusr1.sh belongs to a different PR -- can you please split it out?

https://github.com/lf-edge/eve/pull/1368

File: pkg/pillar/base/execwrapper.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			if start > -1 && end > -1 {
+				file = strings.TrimSpace(filePath[start+1 : end])
+			}
+			break
+		}
+	}
+	switch agent {
+	case "volumemgr", "wstunnelclient", "baseosmgr", "verifier", "downloader", "zedrouter", "zedmanager", "tpmmgr",
+		"vaultmgr", "logmanager", "nim", "ledmanager", "nodeagent", "domainmgr":
+		resultAgentName = agent
```
Comment (eriknordmark) [inline pkg/pillar/base/execwrapper.go:128]: This seems like a huge maintenance pain; can we just have the default case return the agent?

File: pkg/pillar/base/execwrapper.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	case "zedagent":
+		switch file {
```
Comment (eriknordmark) [inline pkg/pillar/base/execwrapper.go:130]: This seems like a smaller maintenance pain; means that any additional touch file in some other agent needs to be known by this code.

But if this is a short-term thing until we can factor out the Exec calls from the main go routine I think we can live with it.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go [hunk 3]
```diff
@@ -354,7 +353,7 @@ func genCredentials() error {
 	err := readCredentials()
 	if err != nil {
 		// Generate a new uuid
-		out, err := exec.Command("uuidgen").Output()
+		out, err := base.ExecCommand(log, false, false, "uuidgen")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:356]: Wait, this should use the uuid package to generate a uuid; no need for an Exec.

File: pkg/pillar/cmd/zedagent/handleinfo.go [hunk 4]
```diff
@@ -0,0 +1,655 @@
+package zedagent
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handleinfo.go:1]: Splitting handlemetrics into two isn't a bad idea, but
1. is better done as a separate PR
2. copyright/license is missing
3. the file should really be called reportinfo.go since it doesn't have any handlers for Info messages.

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 5]
```diff
@@ -445,9 +444,9 @@ func ipRouteCreate(vpnConfig types.VpnConfig) error {
 	if vpnConfig.IsClient {
 		clientConfig := vpnConfig.ClientConfigList[0]
 		tunnelConfig := clientConfig.TunnelConfig
-		cmd := exec.Command("ip", "route", "add", gatewayConfig.SubnetBlock,
-			"dev", tunnelConfig.Name, "metric", tunnelConfig.Metric)
-		if _, err := cmd.Output(); err != nil {
+		if _, err := base.ExecCommand(log, false, false,
+			"ip", "route", "add", gatewayConfig.SubnetBlock, "dev", tunnelConfig.Name,
+			"metric", tunnelConfig.Metric); err != nil {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:449]: Better to fix this to use netlink to add routes as we do elsewhere in the code.

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 6]
```diff
@@ -457,9 +456,8 @@ func ipRouteCreate(vpnConfig types.VpnConfig) error {
 		// for server config, create all client subnet block routes
 		for _, clientConfig := range vpnConfig.ClientConfigList {
 			tunnelConfig := vpnConfig.ClientConfigList[0].TunnelConfig
-			cmd := exec.Command("ip", "route", "add", clientConfig.SubnetBlock,
-				"dev", tunnelConfig.Name, "metric", tunnelConfig.Metric)
-			if _, err := cmd.Output(); err != nil {
+			if _, err := base.ExecCommand(log, false, false, "ip", "route", "add",
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:459]: Ditto

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 7]
```diff
@@ -582,20 +580,18 @@ func ipLinkTunnelCreate(vpnConfig types.VpnConfig) error {
 	log.Infof("%s: %s %s %s\n", tunnelConfig.Name, "ip link add",
 		portConfig.IpAddr, gatewayConfig.IpAddr)
 	if vpnConfig.IsClient {
-		cmd := exec.Command("ip", "link", "add",
+		if _, err := base.ExecCommand(log, false, false, "ip", "link", "add",
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:583]: Better to fix this to use netlink to add links as we do elsewhere in the code.

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 8]
```diff
@@ -605,10 +601,9 @@ func ipLinkTunnelCreate(vpnConfig types.VpnConfig) error {
 	if vpnConfig.VpnRole == AwsVpnClient {
 		log.Infof("%s: %s %s %s\n", tunnelConfig.Name, "ip link addr",
 			tunnelConfig.LocalIpAddr, tunnelConfig.RemoteIpAddr)
-		cmd := exec.Command("ip", "addr", "add",
+		if _, err := base.ExecCommand(log, false, false, "ip", "addr", "add",
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:604]: Ditto

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 9]
```diff
@@ -746,8 +740,7 @@ func ipSecConfigCreate(vpnConfig types.VpnConfig) error {
 	if err := ipSecConfigFileWrite(filename, writeStr); err != nil {
 		return err
 	}
-	cmd := exec.Command("chmod", "600", filename)
-	_, err := cmd.Output()
+	_, err := base.ExecCommand(log, false, false, "chmod", "600", filename)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:743]: Better to use os package instead of Exec

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 10]
```diff
@@ -886,8 +876,7 @@ func checkIntfExistsCmd(intfName string) error {
 }
 
 func checkIntfStateCmd(intfName string) error {
-	cmd := exec.Command("ifconfig", intfName)
-	out, err := cmd.Output()
+	out, err := base.ExecCommand(log, false, false, "ifconfig", intfName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:879]: Netlink should be used instead

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 11]
```diff
@@ -944,8 +932,7 @@ func getIpSecLineState(outArr []string) error {
 }
 
 func issueIfUpCmd(tunnelName string) error {
-	cmd := exec.Command("ifup", tunnelName)
-	_, err := cmd.Output()
+	_, err := base.ExecCommand(log, false, false, "ifup", tunnelName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/ipsec.go:935]: Isn't this just an ip link up, which can be done using netlink?

File: pkg/pillar/base/execwrapper.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+)
+
+const (
+	// Will wait for 1000s max for a call instead of waiting forever
+	maxTimeOutLimit = 1000
+	timeOutLimit    = 100
+)
+
+//ExecCommand executes the given command. Times out if timeOut = true.
+// Also if the exec is taking long, it will update the caller agent's touch file.
```
Comment (rvs) [inline pkg/pillar/base/execwrapper.go:20]: Please include descriptions of each argument in the comment section

File: pkg/pillar/base/execwrapper.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+func ExecCommand(log *LogObject, combined bool, timeOut bool, command string, arg ...string) ([]byte, error) {
```
Comment (rvs) [inline pkg/pillar/base/execwrapper.go:21]: This may be a somewhat controversial opinion, but I really find it difficult to remember what these 2nd and 3d argument stand for. So... would it make sense to have syntactic sugar on top called ExecCommandForever(), etc.

And for the combined use the .CombinedOutput() approach?

WDYT @eriknordmark ?

File: pkg/pillar/cmd/ledmanager/ledmanager.go [hunk 14]
```diff
@@ -452,9 +451,8 @@ func ExecuteDDCmd(ledName string) {
 }
 
 func doDD(count int) {
-	cmd := exec.Command("dd", "if="+diskDevice, "of=/dev/null", "bs=4M",
-		fmt.Sprintf("count=%d", count), "iflag=nocache")
-	stdout, err := cmd.Output()
+	stdout, err := base.ExecCommand(log, false, false,
+		"dd", "if="+diskDevice, "of=/dev/null", "bs=4M", fmt.Sprintf("count=%d", count), "iflag=nocache")
```
Comment (rvs) [inline pkg/pillar/cmd/ledmanager/ledmanager.go:455]: Hey @eriknordmark -- remind me if there was any reason to use dd here as opposed to simply writing to that file from native go io library?

File: pkg/pillar/cmd/zedagent/handleinfo.go (hunk 4 above)
Comment (rvs) [reply pkg/pillar/cmd/zedagent/handleinfo.go:1]: +1 for a separate PR

File: pkg/pillar/cmd/zedrouter/ipsec.go [hunk 15]
```diff
@@ -794,8 +787,7 @@ func ipSecSecretConfigCreate(vpnConfig types.VpnConfig) error {
 	if err := ipSecConfigFileWrite(filename, writeStr); err != nil {
 		return err
 	}
-	cmd := exec.Command("chmod", "600", filename)
-	_, err := cmd.Output()
+	_, err := base.ExecCommand(log, false, false, "chmod", "600", filename)
```
Comment (rvs) [inline pkg/pillar/cmd/zedrouter/ipsec.go:790]: Same as above

File: pkg/pillar/base/execwrapper.go (hunk 1 above)
Comment (adarsh-zededa) [reply pkg/pillar/base/execwrapper.go:128]: In that case, this method will return agentNames which doesn't have touch file to be updated. Example diag and hardwaremodel.

Comment (adarsh-zededa) [reply pkg/pillar/base/execwrapper.go]: Done.

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go (hunk 3 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: Done.

File: pkg/pillar/cmd/zedagent/handleinfo.go (hunk 4 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedagent/handleinfo.go]: https://github.com/lf-edge/eve/pull/1377

File: pkg/pillar/cmd/zedrouter/ipsec.go (hunk 9 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedrouter/ipsec.go]: Done.

File: pkg/pillar/base/execwrapper.go (hunk 13 above)
Comment (adarsh-zededa) [reply pkg/pillar/base/execwrapper.go]: Done.

File: pkg/pillar/cmd/zedrouter/ipsec.go (hunk 15 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedrouter/ipsec.go]: Done.

File: pkg/pillar/cmd/zedrouter/ipsec.go (hunk 5 above)
Comment (adarsh-zededa) [reply pkg/pillar/cmd/zedrouter/ipsec.go]: As per our discussion, opened a tracker to handles this separately.

File: pkg/pillar/base/execwrapper.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
+package base
```
Comment (eriknordmark) [inline pkg/pillar/base/execwrapper.go:1]: Please add a the Copyright+License two lines.

File: pkg/pillar/base/execwrapper.go [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+	case "zedagent":
+		switch file {
+		case "handlecertconfig.go":
+			if methodName == "getCertsFromController" || methodName == "parseControllerCerts" {
+				resultAgentName = "zedagentccerts"
+			} else {
+				resultAgentName = "zedagentattest"
+			}
+		case "handleinfo.go":
+			resultAgentName = "zedagentdevinfo"
```
Comment (eriknordmark) [inline pkg/pillar/base/execwrapper.go:197]: Will this need to change?

Comment (eriknordmark) [reply pkg/pillar/base/execwrapper.go:197]: Is it better to only use the methodName for zedagent i.e., not use the file?

File: pkg/pillar/base/execwrapper.go (hunk 16 above)
Comment (adarsh-zededa) [reply pkg/pillar/base/execwrapper.go:6]: Done.

File: pkg/pillar/base/execwrapper.go (hunk 17 above)
Comment (adarsh-zededa) [reply pkg/pillar/base/execwrapper.go]: >Will this need to change?

Done.

>Is it better to only use the methodName for zedagent i.e., not use the file?

That might lead to having multiple switch cases to figure out an agentName. By having filename, I thought in future even if we add a method to a file which does an exec call, then we don't need to touch this code.

**Review by eriknordmark — COMMENTED** [review]
LGTM. One important nit sent on slack.

File: pkg/pillar/cmd/ledmanager/ledmanager.go (hunk 14 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/ledmanager/ledmanager.go]: @rvs We don't want to write since that causes wear. Thus we read. But normally reads are satisfied by the cache, Hence the nocache flag. I guess we could figure out what dd does with the noocache flag and avoid dd.

Comment (rvs) [reply pkg/pillar/cmd/ledmanager/ledmanager.go]: >  I guess we could figure out what dd does with the noocache flag and avoid dd.

That's exactly my point @eriknordmark -- and here's what it does ;-)

```
fadvise64(0, 0, size of buffer, POSIX_FADV_DONTNEED) = 0
```

So a combination of:

```
unix.Fadvise(..., 4) // POSIX_FADV_DONTNEED
syscall.Madvise(..., 4) // MADV_DONTNEED
```

Should be able to replace this nasty dd invocation.

@adarsh-zededa -- are you up to doing this extra change?

File: pkg/pillar/base/execwrapper.go (hunk 17 above)
Comment (eriknordmark) [reply pkg/pillar/base/execwrapper.go]: The problem is that you have more cases when things will silently break. So we need to limit any refactoring to zedagent until we have moved out the exec calls.

**Review by eriknordmark — APPROVED** [review]
Thanks
